### ISSUE TITLE
TENSORFLOW quantize_graph.py throws error : Graph_def is invalid at node error

### ISSUE THREAD

**Issue by Raj-08** [tensorflow/tensorflow#22238]
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:NA
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:14.04
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**:NA
- **TensorFlow installed from (source or binary)**:PIP and BAZEL
- **TensorFlow version (use command below)**: 1.10
- **Python version**: 2.7
- **Bazel version (if compiling from source)**: 0.15
- **GCC/Compiler version (if compiling from source)**: NA
- **CUDA/cuDNN version**: 9.0
- **GPU model and memory**: NVIDIA 12GB

ERROR ::: Graph_def is invalid at node u'ExpandDims': Input tensor 'image_ph:0' Cannot convert a tensor of type float32 to an input of type int32

I have trained a model , froze it and performed optimization (https://petewarden.com/2016/09/27/tensorflow-for-mobile-poets/) , graph transformation (https://github.com/tensorflow/tensorflow/tree/master/tensorflow/tools/graph_transforms) on tensorflow 1.4.0 till the point which there is no error but when i quantize the model i run into the above mentioned error.

As per https://github.com/tensorflow/tensorflow/issues/4044 i have tried maintaining version consistency which has failed to solve the issue.

**Comment by wt-huang** [comment]
Have you tried the latest TensorFlow version?

**Comment by Raj-08** [comment]
@wt-huang you want me to use latest tensorflow version for quantization only right?
Do I need to train it from scratch and do all of the rest of the process on the latest version ?

**Comment by wt-huang** [comment]
@Raj-08 you would need to use the latest TensorFlow version to train the model, freeze the graph, quantization etc.

**Comment by Raj-08** [comment]
@wt-huang Same problem persists , i have tried all the processes on tensorflow 1.10
"Input 0 of node ExpandDims was passed float from image_ph incompatible with expected int32"

**Comment by wt-huang** [comment]
@Raj-08 I have gone through the steps of model training, freezing the graph and quantization without any issues. Can you list the commands you used also update the above spec sheet with the versions you used?

**Comment by Raj-08** [comment]
I am using default commands given by tensorflow:

`bazel build tensorflow/python/tools:optimize_for_inference
bazel-bin/tensorflow/python/tools/optimize_for_inference \
--input=/tf_files/retrained_graph.pb \
--output=/tf_files/optimized_graph.pb \
--input_names=Mul \
--output_names=final_result`


`bazel build tensorflow/tools/quantization:quantize_graph
bazel-bin/tensorflow/tools/quantization/quantize_graph \
--input=/tf_files/optimized_graph.pb \
--output=/tf_files/rounded_graph.pb \
--output_node_names=final_result \
--mode=weights_rounded`

`bazel build tensorflow/tools/graph_transforms:transform_graph
bazel-bin/tensorflow/tools/graph_transforms/transform_graph \
--in_graph=tensorflow_inception_graph.pb \
--out_graph=optimized_inception_graph.pb \
--inputs='Mul:0' \
--outputs='softmax:0' \
--transforms='
strip_unused_nodes(type=float, shape="1,299,299,3")
remove_nodes(op=Identity, op=CheckNumerics)
fold_old_batch_norms
'`

For freezing i am using my custom script --
With tf.graph.default() as graph:
`input_graph_def = graph.as_graph_def()
    output_node_names =_OUTPUT_NAME
    output_graph_name =  '/home/ubuntu/raj/freeze_inference_graph_mobilenet.pb'

    with tf.Session() as sess:
        saver = tf.train.import_meta_graph('/home/ubuntu/raj/research/train_mat_dp_1/model.ckpt-0.meta',clear_devices=True)
        saver.restore(sess, '/home/ubuntu/raj/research/train_mat_dp_1/model.ckpt-0')
        print ("Exporting graph...")
        output_graph_def = graph_util.convert_variables_to_constants(
            sess,
            input_graph_def,
           [_OUTPUT_NAME])

        with tf.gfile.GFile(output_graph_name, "wb") as f:
            f.write(output_graph_def.SerializeToString())`

**Comment by wt-huang** [comment]
@Raj-08 could you check whether there are any formatting issues in your model transformation. Make sure the inputs, outputs and weights are correct depending upon the mobile device you are using.

**Comment by Raj-08** [comment]
One thing which i want to ask is this , i am using tensorflow inference on android , now i have been using two ops tf.nn.dilation and tf.nn.erosion in my training , and while inference it says no op kernel was registered for this , now is it possible that tensorflow inference on android is not able to understand these ops even though these ops were working on tensorflow GPU?

**Comment by wt-huang** [comment]
@Raj-08 Currently tf.nn.dilation and tf.nn.erosion are not supported in TFLite per se, some of the dilation/erosion options are embedded in conv that are potentially modifiable. Alternatively, you can add a custom op as described [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/lite/g3doc/custom_operators.md) or only use the ops the are currently supported by TFLite in your model.

**Comment by wt-huang** [comment]
Closing this issue, feel free to open a new issue if any error comes up.
